Add waitNext() to BlockTemplate interface

Sjors opened the pull request:
This PR introduces `waitNext()`. It waits for either the tip to update or for fees at the top of the mempool to rise sufficiently. It then returns a new template, with which the caller can rinse and repeat.

On testnet3 and testnet4 the difficulty drops after 20 minutes, so the second ensures that a new template is returned in that case.

Alternative approach to #31003, suggested in https://github.com/bitcoin/bitcoin/issues/31109#issuecomment-2451942362

Sjors commented on src/node/interfaces.cpp:
```diff
@@ -938,9 +938,71 @@ class BlockTemplateImpl : public BlockTemplate
+        return block_template;
+    }
+
+    const CScript m_script_pub_key;
```
Separate from this PR, I'm looking into whether we can avoid passing `script_pub_key` around. Afaik it's only used by tests and for solo CPU mining.

Sjors commented on src/node/interfaces.cpp:
```diff
@@ -938,9 +938,71 @@ class BlockTemplateImpl : public BlockTemplate
 
+    std::unique_ptr<BlockTemplate> waitNext(CAmount fee_threshold, MillisecondsDouble timeout) override
+    {
+        if (timeout > std::chrono::years{100}) timeout = std::chrono::years{100}; // Upper bound to avoid UB in std::chrono
```
I still need to study https://github.com/bitcoin/bitcoin/pull/31003#discussion_r1792221843 in more detail to see if this code isn't making similar mistakes.

Sjors replied:
I dropped this line.

Additionally this PR does not make use of `last_mempool_update`, since it's not very useful in a typical production setting where something in the mempool updates all the time.

I think the race condition mentioned there is not an issue here because we don't pass in the current fees via argument, but instead derive them from the original template.

I did forget to update `now` at the end of the loop, which I fixed.

Sjors commented:
I pushed a commit refactoring `miner_tests` to use the `Mining` interface. I plan to expand those tests to cover `waitNext()`.

Sjors commented on src/test/miner_tests.cpp:
```diff
@@ -54,23 +56,17 @@ struct MinerTestingSetup : public TestingSetup {
-    BlockAssembler::Options options;
-
-    options.nBlockMaxWeight = MAX_BLOCK_WEIGHT;
-    options.blockMinFeeRate = blockMinFeeRate;
```
a316ba58cba5df8aadb05737959b97eeb6a350b4: I think dropping this doesn't matter?

Sjors commented on src/test/miner_tests.cpp:
```diff
@@ -589,59 +605,66 @@ void MinerTestingSetup::TestPrioritisedMining(const CScript& scriptPubKey, const
-        BOOST_CHECK(Assert(m_node.chainman)->ProcessNewBlock(shared_pblock, true, true, nullptr));
-        pblock->hashPrevBlock = pblock->GetHash();
+        std::shared_ptr<const CBlock> shared_pblock = std::make_shared<const CBlock>(block);
+        BOOST_REQUIRE(Assert(m_node.chainman)->ProcessNewBlock(shared_pblock, /*force_processing=*/true, /*min_pow_checked=*/true, nullptr));
```
a316ba58cba5df8aadb05737959b97eeb6a350b4: I'm tempted to use `miner.submitSolution()` here.

Sjors replied:
Done

Sjors commented:
I made the test code alternate calls between Chainman's `ProcessNewBlock` and `submitSolution` via the Mining interface.

TheCharlatan commented:
In commit f9200f6c65b5a97a4842cc1ec34e688ec13bffcd

"Use createNewBlock via the interface instead of calling Chainman's CreateNewBlock."

Do you mean "Use createNewBlock via the interface instead of calling CreateNewBlock through the BlockAssembler directly"? There is no `CreateNewBlock` in the chainman.

Sjors commented:
Made some adjustments based on https://github.com/bitcoin/bitcoin/pull/31283#discussion_r1842225267

Sjors commented:
@TheCharlatan I adjusted the commit message

Sjors commented on src/node/interfaces.cpp:
```diff
@@ -938,9 +938,76 @@ class BlockTemplateImpl : public BlockTemplate
+                    // Skip coinbase
+                    if (fee < 0) continue;
+                    new_fees += fee;
+                    if (new_fees >= current_fees + fee_threshold) return block_template;
```
A `fee_threshold` of 0 combined with a `timeout` of 0 will immediately return a new template. This is useful for the Template Provider, because it needs to unconditionally send new templates to all connected clients if fees increased sufficiently for one its connected clients. (at least until cluster mempool makes the fee calculation trivially cheap)

Sjors commented:
Added early return logic. Similar to #31297 this would be less tedious if `notifications().m_tip_block` was set the tip earlier.

Sjors commented:
Marking this draft, let's review the early return and m_tip_block block changes in #31297 first.

Sjors commented:
#31297 is on hold, as it's probably unnecessary.

Please review #31325 first, so I can decide whether to build this PR on top of it.

Sjors commented:
Rebased after #31288 and #31122.

This PR is now based on #31325 since that has enough support. Keeping this draft until that's merged (or dropped).

The test changes will probably cause a merge conflict with #31318, but it should be trivial.

I dropped the early return in `waitNext()` because `wait_until` should already take care of that. I also made the while loop use `<=` instead of `<` so this actually works: https://github.com/bitcoin/bitcoin/pull/31283#discussion_r1843589729 Added a test for that.

Sjors commented:
Fixed a bug found while testing https://github.com/Sjors/bitcoin/pull/49, where it was spuriously making new block templates every tick even if fees didn't rise. [diff](
https://github.com/bitcoin/bitcoin/compare/245660284de8b4ac29428737f088c22cb2812037..5d294d86f43c83603b5d965eff36c90159c6bffe)

Sjors commented on src/node/interfaces.cpp:
```diff
@@ -938,9 +938,85 @@ class BlockTemplateImpl : public BlockTemplate
+                    // Although C++17 allows safe comparison of optional<T> with
+                    // another T, we need to wait for m_tip_block to be set AND
+                    // for the value to be different than the current_tip value.
+                    tip_changed = notifications().m_tip_block && notifications().m_tip_block != m_block_template->block.hashPrevBlock;
```
If #31346 lands first, `notifications().m_tip_block` could be an `Assert` and the comment removed, see https://github.com/bitcoin/bitcoin/pull/31346#issuecomment-2493956840

Sjors commented:
Rebased after #31112.

Sjors commented:
Rebased after https://github.com/bitcoin/bitcoin/pull/31346.

Sjors commented:
Rebased after https://github.com/bitcoin/bitcoin/pull/31318 and #31197.

Sjors committed:
Add waitNext() to BlockTemplate interface

Sjors commented:
Fixed tidy error.

Sjors commented:
The changes here don't conflict with the merged #31196, but I rebased just in case.

Sjors commented:
#31325 landed so this is ready for review.

Sjors commented:
I extracted the first commit into its own PR #31581. Review on the second commit is still welcome here, since I don't expect it to change. Marking draft.

Sjors commented:
Rebased on the latest #31581.

I added a commit that on testnet3 and testnet4 returns a new template after 20 minutes, to take advantage of the min difficulty rule.

Since `miner_tests.cpp` uses mainnet, I created `testnet4_miner_tests.cpp` to cover this behaviour in the unit tests. A functional test could be added later using the testnet4 blocks added by #31583.

I also switched to using `NodeClock::now()` to be able to use mock time in tests.

Sjors commented:
Rebased after #31581 landed, marking ready for review again.

ryanofsky reviewed (approved):
Code review ACK 8d54cfdb1f368bd4486bbc4b285f4712a0f260fa, but I didn't look at the tests very closely and could review them more. The code seems to have gotten simpler since the last version in #31003, so that is very nice

Sjors commented on src/node/interfaces.cpp:
```diff
@@ -936,9 +936,87 @@ class BlockTemplateImpl : public BlockTemplate
+            }
+
+            // Break out of while when using mock time
+            if (now == deadline) break;
```
Even after dropping the `fee_threshold==0 && timeout==0` special case, I still need this. Added a comment to explain why (unit tests can't easily move mock time during a blocking call).

vasild replied:
Just stumbled on this and posted a [comment to the PR elsewhere](https://github.com/bitcoin/bitcoin/pull/31283#discussion_r1937133582)... Having this only to make unit tests easier seems not enough justification to me.

vasild replied:
> unit tests can't easily move mock time during a blocking call

Consider the following pattern to do that. I think it would be better to have that in the tests instead of this `if (now == deadline)` in the main (production) code.

```cpp
std::atomic_int now = 0;

void waitNext(int timeout)
{
    const int deadline = now + timeout;
    while (now < deadline) {
        std::cout << "waitNext() now=" << now << "\n";
        std::this_thread::sleep_for(500ms);
    }
}

int main(int, char**)
{
    std::atomic_bool stop_advancing_mocktime{false};
    std::thread advance_mocktime{[&stop_advancing_mocktime]() {
        while (!stop_advancing_mocktime) {
            ++now;
            std::this_thread::sleep_for(500ms);
        }
    }};
    waitNext(5);
    stop_advancing_mocktime = true;
    advance_mocktime.join();
    return 0;
}
```

ryanofsky replied:
> I think it would be better to have that in the tests instead of this `if (now == deadline)` in the main (production) code.

Do we actually think that behavior is bad to have in production code? It seems like it's perfectly reasonable for production code to return either at the deadline or immediately after, and the choice is basically arbitrary for normal code but important for tests, so we are choosing a behavior that makes it more convenient to write tests and documenting that fact. This doesn't seem l like a bad thing to me, and it might be preferable to introducing another thread and risking making the tests nondeterministic. Just my thought, no strong opinion though.

vasild replied:
Functionality-wise I think the behavior of the production code is ok. I am concerned about its readability and maintainability. It just looks confusing (to me).

Sjors commented:
I dropped the special case for `fee_threshold` of 0 combined with a `timeout` of 0 and the related mock time workaround. I'll adjust https://github.com/Sjors/bitcoin/pull/49.

I also switched to using a new `BlockWaitOptions` struct.

Also rebased.

ryanofsky reviewed (approved):
Code review ACK 6bf77e5a376804c0681897fa032537ba16e9a99a. Code looks great, but I still would like to spend more time reviewing tests in detail.

Main changes since last review were just adding options struct, dropping special case, adding comment explaining the now == deadline logic and other documentation. Code is pretty clear and simple now without the mempool GetTransactionsUpdated complexity, without the previous special case, and with the updated documentation.

ismaelsadeeq commented on src/interfaces/mining.h:
```diff
@@ -56,6 +56,16 @@ class BlockTemplate
+    /**
+     * Waits for fees in the next block to rise, a new tip or the timeout.
+     *
+     * @param[in] options   Control the timeout (default forever) and by how much total fees
+     *                      for the next block should rise (default infinite).
```
In  "Add waitNext() to BlockTemplate interface" a61fb2fd2d6b338c08c619e4605521a41bd3edd9

We already have the comments about the default in node types `BlockWaitOptions` docstring, writing it here is duplicating same information, when we update the default we have to update both places.

I think it will be better to just move the fee threshold description comment to node types `BlockWaitOptions` docstring and delete the duplicate here.

Sjors replied:
I tried to make it a quick TL&DR so someone can understand the most important behaviour, without having to go and read the entire struct doc (which might grow, and not all code let you hover and see the full thing in a popup).

Sjors replied:
(and these two defaults are very unlikely to change).

ismaelsadeeq commented on src/node/interfaces.cpp:
```diff
@@ -936,9 +937,88 @@ class BlockTemplateImpl : public BlockTemplate
+             * is to generate a fresh template. Cluster Mempool may allow for a more efficient
+             * way to determine how much (approximate) fees for the next block increased.
+             *
+             * We'll also create a new template if the tip changed during the last tick.
+             */
+            if (options.fee_threshold < MAX_MONEY || tip_changed) {
+                auto block_template{std::make_unique<BlockTemplateImpl>(m_assemble_options, BlockAssembler{chainman().ActiveChainstate(), context()->mempool.get(), m_assemble_options}.CreateNewBlock(), m_node)};
```
In "Add waitNext() to BlockTemplate interface" a61fb2fd2d6b338c08c619e4605521a41bd3edd9 

I think we are not just creating a block template after ticks but on each iteration of this loop, which continues infinitely in the default case. Since we are holding `cs_main` continuously, wouldn't this cause resource contention and delay other components of the node?

My concern is that, regardless of whether the mempool improves, we keep creating block templates.

Currently there is no reliable way to determine if the fee rate diagram of the mempool has increased, I think this will become easier after #30289. When RBF occurs, we can be certain that the fee rate diagram of the mempool has improved. We could even determine the fee additions after those RBFs and compare them with `fee_threshold` the pool want; if the threshold is exceeded, we should trigger a block template update.

Since that is not possible and #30289  is a WIP I think we should only build a block template after some time has passed. It seems like GBT waits for a minute maybe you could do the same?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/31283#discussion_r1936237852

> In "Add waitNext() to BlockTemplate interface" [a61fb2f](https://github.com/bitcoin/bitcoin/commit/a61fb2fd2d6b338c08c619e4605521a41bd3edd9)
> 
> I think we are not just creating a block template after ticks but on each iteration of this loop, which continues infinitely in the default case. Since we are holding `cs_main` continuously, wouldn't this cause resource contention and delay other components of the node?

The `m_tip_block_cv.wait_until(lock, std::min(now + tick, deadline))` call on [line 957](https://github.com/bitcoin/bitcoin/blob/6bf77e5a376804c0681897fa032537ba16e9a99a/src/node/interfaces.cpp#L957C1-L957C24) is what is supposed to prevent this. Each time that line is reached the thread should sleep for one tick. It should only wake up before the tick if the tip changes or the call is timing out.

Sjors replied:
Did you mean a61fb2fd2d6b338c08c619e4605521a41bd3edd9?

`cs_main` is only held briefly, _after_ each 1 second wait, once per loop.

This PR currently makes 1 template per second to get the fees. Once we have cluster mempool, I'm hoping there will be a lighter weight method to see fees in the top 4 mega-weight units have risen. cc @sipa / @sdaftuar 

One of the stratum v2 goals is to push new templates faster, so I'd rather keep the tick at 1 second unless it causes problems on slow hardware.

ismaelsadeeq commented on src/node/interfaces.cpp:
```diff
@@ -936,9 +937,88 @@ class BlockTemplateImpl : public BlockTemplate
+                }
+
+                // Calculate the original template total fees if we haven't already
+                if (current_fees == -1) {
+                    current_fees = 0;
+                    for (CAmount fee : m_block_template->vTxFees) {
+                        // Skip coinbase
+                        if (fee < 0) continue;
+                        current_fees += fee;
+                    }
+                }
+
+                CAmount new_fees = 0;
+                for (CAmount fee : block_template->m_block_template->vTxFees) {
+                    // Skip coinbase
+                    if (fee < 0) continue;
+                    new_fees += fee;
+                    if (new_fees >= current_fees + options.fee_threshold) return block_template;
+                }
```
In "Add waitNext() to BlockTemplate interface"  a61fb2fd2d6b338c08c619e4605521a41bd3edd9

@Sjors  IIRC block assembler has a total fees field `nFees` that also skips coinbase transaction, why aren't we using that? could expose it and use it, this will prevent the iterating through all the block transactions to compute the total fees during each tick?

Sjors replied:
Good point. `BlockAssembler::CreateNewBlock()` uses `nFees` internally. It's updated in `AddToBlock`. I could expose it via `CBlockTemplate`.

However this code will hopefully go away soon(tm) with Cluster Mempool, and I suspect it's extremely fast. So it's perhaps not worth refactoring. In any case it could be done in a followup without changing the interface.

ismaelsadeeq commented on src/interfaces/mining.h:
```diff
@@ -64,6 +64,9 @@ class BlockTemplate
      *                      for the next block should rise (default infinite).
```
In "On testnet create min diff template after 20 mins" 6bf77e5a376804c0681897fa032537ba16e9a99a

nit: 
@Sjors should this commit message and comments be updated to reflect the actual change something like

"miner: update `waitNext` to return a new block if 20 minutes have elapsed since the tip"

I think the above discription is exactly what this commit does and does not enable the creation of the min diff template, it's done internally by the block assembler?

Sjors replied:
Changed the commit message.

ismaelsadeeq reviewed (commented):
Code Review 6bf77e5a376804c0681897fa032537ba16e9a99a

Just a question and a nit

vasild commented on src/node/interfaces.cpp:
```diff
@@ -936,9 +937,99 @@ class BlockTemplateImpl : public BlockTemplate
+                }
+
+                // Calculate the original template total fees if we haven't already
+                if (current_fees == -1) {
+                    current_fees = 0;
+                    for (CAmount fee : m_block_template->vTxFees) {
+                        // Skip coinbase
+                        if (fee < 0) continue;
+                        current_fees += fee;
+                    }
+                }
```
Looking at

https://github.com/bitcoin/bitcoin/blob/8fa10edcd1706a1f0dc9d8c3adbc8efa3c7755bf/src/node/miner.cpp#L157

isn't this equivalent to just `current_fees = m_block_template->vTxFees[0];`?

Sjors replied:
That might work actually, with an extra minus. But it seems brittle, because the line `pblocktemplate->vTxFees[0] = -nFees;` doesn't do anything afaik and might one day get deleted, causes this code to thing fees are 0.

Sjors replied:
I'd rather remove the line, it might have had a use before that no longer applies.

If we want to get fees via the CBLockTemplate then it should just get a field for that.

But I don't think it's useful to do any of that, because this (duplicate) code will likely go away soon: https://github.com/bitcoin/bitcoin/pull/31283#discussion_r1937451223

vasild commented on src/node/interfaces.cpp:
```diff
@@ -936,9 +937,99 @@ class BlockTemplateImpl : public BlockTemplate
+            // On test networks return a minimum difficulty block after 20 minutes
+            if (!tip_changed && allow_min_difficulty) {
+                const NodeClock::time_point tip_time{std::chrono::seconds{chainman().ActiveChain().Tip()->GetBlockTime()}};
+                if (now > tip_time + std::chrono::seconds(20 * 60)) {
```
nit:
```suggestion
                if (now > tip_time + 20min) {
```

Sjors replied:
I think I lost it in a rebase, taken again.

vasild reviewed (commented):
Approach ACK 6bf77e5a37

vasild commented on src/node/interfaces.cpp:
```diff
@@ -936,9 +937,99 @@ class BlockTemplateImpl : public BlockTemplate
+            }
+
+            /**
+             * Break out of while when using mock time. While functional tests
+             * can call setmocktime to move the clock and exit this loop, a
+             * unit tests would need to spawn a thread to achieve this.
+             * Making the while loop use now < deadline won't work either, because
+             * then there's no wait to test its internals, e.g. the  20 minute
+             * testnet exception.
+             */
+            if (now == deadline) break;
```
Lets continue the discussion here https://github.com/bitcoin/bitcoin/pull/31283#discussion_r1925533767 which is a previous comment about the same thing.

Sjors replied:
If the loop uses `while (now < deadline)` then it will never generate a template, so all the `BOOST_REQUIRE(block_template)` checks will fail.

So it has to use `<=`, but without the `break` here the test will freeze, because time doesn't move.

vasild replied:
You mean "never generate a template" when run from the tests, right? Why is that? The default timeout is `inf` so `now < deadline` should be `true`?

> So it has to use <=, but without the break here the test will freeze, because time doesn't move.

Right, but it is not too difficult to advance the time in unit tests.

Sjors replied:
It would require a second thread, because the main thread of the unit test is waiting on `waitNext()`.

vasild replied:
Ok, it will never generate a template because it will never exit from `waitNext()` (when run from unit tests with mocktime which never advances). "BOOST_REQUIRE(block_template) checks will fail" -- they will never be reached in that case.

What about this (unit tests will use timeout=0):

```
now = now()
deadline = now + timeout;
do {
    stuff;
    now = now();
} while (now < deadline);
```
it will execute the body of the loop exactly once when given timeout=0 regardless of whether the clock advances or not.

Sjors replied:
I think it's better explicitly mention the test issue:
1. Less likely for someone to accidentally break it and be confused
2. Maybe someone adds a nice macro one day `RUN_COMMAND_DO_OTHER_THING_IN_NEW_THREAD_AFTER(command, delay)`

vasild replied:
> then there's no wait to test its internals

Should that be "no way"?

Sjors replied:
Will push fix if I need to change anything else.

vasild replied:
Consider the following. It avoids the hackish part and achieves the same purpose:

```diff
diff --git i/src/node/interfaces.cpp w/src/node/interfaces.cpp
index 9fe4984cd0..8022b47b5e 100644
--- i/src/node/interfaces.cpp
+++ w/src/node/interfaces.cpp
@@ -960,13 +960,13 @@ public:
             // could have been generated before a tip exists.
             Assume(tip_block);
             tip_changed = tip_block != m_block_template->block.hashPrevBlock;
             return tip_changed;
         };
 
-        while (now <= deadline) {
+        do {
             {
                 WAIT_LOCK(notifications().m_tip_block_mutex, lock);
                 // Note that wait_until() checks the predicate before waiting
                 notifications().m_tip_block_cv.wait_until(lock, std::min(now + tick, deadline), [&]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
                     return check_tip_changed() || chainman().m_interrupt;
                 });
@@ -1018,23 +1018,14 @@ public:
                     new_fees += fee;
                     Assume(options.fee_threshold != MAX_MONEY);
                     if (new_fees >= current_fees + options.fee_threshold) return block_template;
                 }
             }
 
-            /**
-             * Break out of while when using mock time. While functional tests
-             * can call setmocktime to move the clock and exit this loop, a
-             * unit tests would need to spawn a thread to achieve this.
-             * Making the while loop use now < deadline won't work either, because
-             * then there's no wait to test its internals, e.g. the  20 minute
-             * testnet exception.
-             */
-            if (now == deadline) break;
             now = NodeClock::now();
-        }
+        } while (now < deadline);
 
         return nullptr;
     }
 
     const BlockAssembler::Options m_assemble_options;
 
```

Sjors replied:
Taken

Sjors commented on src/node/types.h:
```diff
@@ -61,6 +63,22 @@ struct BlockCreateOptions {
     CScript coinbase_output_script{CScript() << OP_TRUE};
 };
+
+struct BlockWaitOptions {
+    /**
+     * How long to wait before returning nullptr instead of a new template.
+     * Default is to wait forever.
+     */
+    MillisecondsDouble timeout{MillisecondsDouble::max()};
+
+    /**
+     * Wait until total fees in the new template exceed fees in the origal
+     * template by at least this amount (in sats). The default is to ignore
+     * fee increases and only wait for a tip change.
+     */
+    CAmount fee_threshold{MAX_MONEY};
```
IIRC I used magic values back when these were passed in as arguments, but in a struct using an `std::optional` is not much friction. I'll look into that.

ryanofsky replied:
> IIRC I used magic values back when these were passed in as arguments, but in a struct using an `std::optional` is not much friction. I'll look into that.

IMO, MAX_MONEY is a magic value that would be good to replace with nullopt, but MillisecondsDouble::max() is not a magic value and the code would be simpler if it were kept.

A magic value (imo) is a value that is handled specially and adds a special case, so replacing `CAmount fee_threshold` with std::optional` is good because the code actually has a *contains* a special case for not checking for any fee threshold, and so it is better to make that special case more explicit and directly controllable.

But the code does not contain any special case for MillisecondsDouble::max(), it is just treated correctly as a really large value, so changing that to use std::optional would only add a special case that wasn't previously there and make the code more complicated and fragile. So it would be better to keep the way it is.

vasild replied:
Agreed about `MillisecondsDouble::max()`. Note here that we rely to have such a max value that when we add to it the result is the same max value (not overflow or cause unexpected things).

Sjors replied:
@ryanofsky do I understand correctly that for the `.capnp` there's no difference when making something an `std::optional`?

ryanofsky replied:
> @ryanofsky do I understand correctly that for the `.capnp` there's no difference when making something an `std::optional`?

It depends on the type. In capnproto schemas primitive fields (numbers, bools, enums, unions) are not optional and are always present. If these fields are not assigned values they are just initialized to 0. So to represent a std::optional\<T> value where T is a primitive you have to change the capnproto schema to be able to represent std::nullopt differently somehow, unless you are ok with std::nullopt being represented as 0.

Capnproto does allow non-primitive fields (data, text, struct, list fields) to be missing so you don't need to change the capnproto schema to use std::optional for these. For these fields a std::nullopt field is just serialized with the field unset, and this is not ambiguous.

The difference between primitive/non-primitve fields is explained at https://capnproto.org/faq.html#how-do-i-make-a-field-optional and the faq recommends using a union with Void to represent optional primitive fields(*). Unfortunately though, I didn't know about that trick and I don't think libmultiprocess currently supports that, although it could. Another natural way to represent an optional value might just to use a List() type and only populate it with 0 or 1 elements.

But actually the way I've supported optional primitive fields in libmultiprocess so far has just been to accompany them with a boolean "has" field. So for example the `Chain::getHeight` method which returns `std::optional<int>` in c++ returns [`(result :Int32, hasResult :Bool)`](https://github.com/ryanofsky/bitcoin/blob/64b833854a34d87cde4e0ca4173d75012c401a7a/src/ipc/capnp/chain.capnp#L20) in capnproto. I know this works for parameters and returns values, though I'm not sure if it works for struct fields. You could just declare the struct to have a `hasFeeThreshold: Bool` value and I could fix this in libmultiprocess if it isn't working.

Or if we think maybe `List(Int64)` or `Union(Void, Int64)` would be better ways to represent `std::optional<Int64>` fields we could use one of those representations and I can extend libmultiprocess to support them.

(*)

vasild replied:
Ok, I think using `MAX_MONEY` as magic value to mean "no threshold" would be nice to avoid but is not a blocker for this PR. Accompanying boolean "has" field is fine too.

ryanofsky commented on src/node/interfaces.cpp:
```diff
@@ -936,9 +937,99 @@ class BlockTemplateImpl : public BlockTemplate
+        // Alternate waiting for a new tip and checking if fees have risen.
+        // The latter check is expensive so we only run it once per second.
+        auto now{NodeClock::now()};
+        const auto deadline = now + options.timeout;
```
> But if that is changed to an integer type this will break (overflow).

I don't think there is anything unusual about that. If you change a floating point variable to an integer, you should expect code using that variable to need to be updated to handle overflows correctly, because integer and floating point numbers overflow differently. This code could work correctly if it were switched to an integer. It would just need to use saturating integer operations like the ones in util/overflow.h.

I also think an nullopt value is more semantically ambiguous. Specifying nullopt or "no timeout" could mean either return right away or wait forever. But specifying 1.7 x 10^308 seconds more clearly corresponds to waiting forever.

Sjors commented on src/node/interfaces.cpp:
```diff
@@ -936,9 +937,88 @@ class BlockTemplateImpl : public BlockTemplate
+            LOCK(::cs_main);
+            /**
+             * The only way to determine if fees increased compared to the previous template,
+             * is to generate a fresh template. Cluster Mempool may allow for a more efficient
```
I was hoping we could simply have a variable that tracks the total fees for the top 1 vMB, which is constantly updated as things are inserted and removed from the mempool.

ismaelsadeeq replied:
@sipa, I imagine that an RBF over any transaction in the last built block template be an indicator that the previously built template improve.

And since we would be using fee rate diagram for RBF's after cluster mempool, we could use use RBF's as determinant?

Sjors replied:
I adjusted the comment.

Sjors committed:
miner: have waitNext return after 20 min on testnet

On testnet we need to create a min diff template after 20 min.

Sjors commented on src/interfaces/mining.h:
```diff
@@ -56,6 +56,19 @@ class BlockTemplate
+     * @param[in] options   Control the timeout (default forever) and by how much total fees
+     *                      for the next block should rise (default infinite).
+     *
+     * @returns a new BlockTemplate or nullptr if the timeout occurs.
```
Changed to "nothing"

Sjors commented on src/node/interfaces.cpp:
```diff
@@ -936,9 +937,99 @@ class BlockTemplateImpl : public BlockTemplate
+            {
+                WAIT_LOCK(notifications().m_tip_block_mutex, lock);
+                notifications().m_tip_block_cv.wait_until(lock, std::min(now + tick, deadline), [&]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
+                    Assert(notifications().TipBlock());
```
I turned `Assert` into `Assume`. When comparing an std::optional<uint256> with a <uint256> it will be considered unequal if the optional is empty. https://en.cppreference.com/w/cpp/utility/optional/operator_cmp (24)

This would trigger a spurious `tip_changed`. But since it can only happen during shutdown, it's not a problem. And not fatal in any case.

Sjors commented:
Rebased and addressed new feedback.

ryanofsky commented on src/ipc/capnp/mining.capnp:
```diff
@@ -39,6 +40,11 @@ struct BlockCreateOptions $Proxy.wrap("node::BlockCreateOptions") {
     coinbaseOutputMaxAdditionalSigops @2 :UInt64 $Proxy.name("coinbase_output_max_additional_sigops");
 }
 
+struct BlockWaitOptions $Proxy.wrap("node::BlockWaitOptions") {
```
In commit "miner: have waitNext return after 20 min on testnet" (22cf6cdb4190482b486752dae4e9dc3b44cfcb1e):

Note for a possible followup. I think this representation of the options struct should work fine for now as long as the client initializing the BlockWaitOptions struct is a c++ client using the libmultiprocess library.

However, this representation of the options struct might not be the most convenient for other clients, like a possible rust client, because by default these fields will both be set to 0 and interpreted by the server as 0, so the rust client will have to manually remember to set these fields to max values if they want default behavior.

This problem could be avoided by adding additional `hasTimeout` and `hasFeeThreshold` bool fields which will both default to false, and which should cause the c++ server code code to automatically leave the default c++ struct field values in place and not overwrite them with 0. Or other representations like the union(Void, T) representation mentioned https://github.com/bitcoin/bitcoin/pull/31283#discussion_r1937622086 could also work. Any representation that results in default behavior when a capnproto client leaves these fields unset would work.

This can be a followup though, since default libmultiprocess c++ client will initialize both these fields from a source C++ struct which has the correct default values.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/31283#discussion_r1937644528

Actually never mind all of this, I am completely wrong. Capnproto **does** support setting default values, using syntax shown https://capnproto.org/language.html#structs, but for some reason I seemed to have a memory of it not supporting them. (It supports default values using a xor trick described https://capnproto.org/encoding.html#default-values).

So recommendation here could be to set default values in the capnproto struct to match default values in the c++ struct. But again this could be a followup and it only affects other potential clients like rust clients, not c++ client using libmultiprocess since they will just be serializing a c++ struct which already has the defaults set.

Sjors replied:
I'd really like this to be easy to use for a Rust client as early as possible. It seems better to revert to the magic value for now.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/31283/commits/dcb75cbae61918a13310e3e22190c088d9b7605c#r1937644528

> I'd really like this to be easy to use for a Rust client as early as possible. It seems better to revert to the magic value for now.

To make this easy for rust (or other) cilents using this interface you can assign these fields default values. Maybe define MAX_MONEY and MAX_TIMEOUT constants using syntax shown https://capnproto.org/language.html#constants and assign them as defaults using syntax shown https://capnproto.org/language.html#structs. These changes could be a followup though and shouldn't be necessary for this PR.

Sjors commented on src/node/interfaces.cpp:
```diff
@@ -936,9 +937,90 @@ class BlockTemplateImpl : public BlockTemplate
+            {
+                WAIT_LOCK(notifications().m_tip_block_mutex, lock);
+                notifications().m_tip_block_cv.wait_until(lock, std::min(now + tick, deadline), [&]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
+                    Assume(notifications().TipBlock());
```
The assume is correct, but the assert was a bit overzealous. `waitNext()` is a method on `BlockTemplate`. You can't generate a template before there's a tip, so it's impossible for this fail.

It's possible that the node is shutdown when `TipBlock()` still returns `std::nullopt`, but in that scenario there can't be a `BlockTemplate`. The value doesn't get unset during shutdown.

Sjors replied:
A comment to remember this.

vasild replied:
> but in that scenario there can't be a `BlockTemplate` ...

What about

1. A `BlockTemplate` is generated
2. Shutdown is initiated
3. This code is reached and `TipBlock()` returns an empty optional, triggering the `Assume()`

Can't happen?

Sjors replied:
Step (3) doesn't happen. A shutdown doesn't null out `m_tip_block`.

vasild replied:
But what about this comment:

https://github.com/bitcoin/bitcoin/blob/1172bc4157eefe80d1aaf0b56459857ec651e535/src/node/kernel_notifications.h#L61-L62

related:

https://github.com/bitcoin/bitcoin/blob/1172bc4157eefe80d1aaf0b56459857ec651e535/src/init.cpp#L1817-L1827

Sjors replied:
@vasild it's possible to start the node and quickly shut it down. If that happens then `m_tip_block` remains `null`.

If you call `createNewBlock()` during that time I'm not sure what happens.  I think it's best to add an extra guard there. The Template Provider calls `waitTipChanged` before calling `createNewBlock()`, but not every interface user might do that.

With that in place you can't get here.

Sjors replied:
Opened #31785 for this check, though I could include it here.

Sjors replied:
> though I could include it here.

It turned into a bigger change than I expected, so better to keep it separate from this. I don't think it's blocking.

vasild replied:
Or this, idea from [discussion from another thread](https://github.com/bitcoin/bitcoin/pull/31283#discussion_r1937078288) - return `nullptr` if `TipBlock()` returns no block:

<details>
<summary>[patch] return nullptr from waitNext() if no tip</summary>

```diff
diff --git i/src/node/interfaces.cpp w/src/node/interfaces.cpp
index 1eb66d81c0..9af5e65308 100644
--- i/src/node/interfaces.cpp
+++ w/src/node/interfaces.cpp
@@ -947,35 +947,37 @@ public:
         // The latter check is expensive so we only run it once per second.
         auto now{NodeClock::now()};
         const auto deadline = now + options.timeout;
         const MillisecondsDouble tick{1000};
         const bool allow_min_difficulty{chainman().GetParams().GetConsensus().fPowAllowMinDifficultyBlocks};
 
+        std::optional<uint256> tip_hash;
         bool tip_changed{false};
         // Helper to check if the tip has changed, and also update tip_changed.
-        auto check_tip_changed = [this, &tip_changed]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
+        auto check_tip_changed = [this, &tip_hash, &tip_changed]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
             AssertLockHeld(notifications().m_tip_block_mutex);
             if (tip_changed) return true;
-            const auto tip_block{notifications().TipBlock()};
-            // This is an instance method on BlockTemplate and no template
-            // could have been generated before a tip exists.
-            Assume(tip_block);
-            tip_changed = tip_block != m_block_template->block.hashPrevBlock;
+            tip_hash = notifications().TipBlock();
+            if (tip_hash.has_value()) {
+                tip_changed = tip_hash.value() != m_block_template->block.hashPrevBlock;
+            } else {
+                tip_changed = true;
+            }
             return tip_changed;
         };
 
         do {
             {
                 WAIT_LOCK(notifications().m_tip_block_mutex, lock);
                 // Note that wait_until() checks the predicate before waiting
                 notifications().m_tip_block_cv.wait_until(lock, std::min(now + tick, deadline), [&]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
                     return check_tip_changed() || chainman().m_interrupt;
                 });
             }
 
-            if (chainman().m_interrupt) return nullptr;
+            if (chainman().m_interrupt || !tip_hash.has_value()) return nullptr;
 
             // Must release m_tip_block_mutex before locking cs_main, to avoid deadlocks.
             LOCK(::cs_main);
 
             // On test networks return a minimum difficulty block after 20 minutes
             if (!tip_changed && allow_min_difficulty) {
```
</details>

Does that look sensible?

Sjors replied:
I think #31785 is a more thorough solution that avoids having to deal with this edge case all over the place.

Sjors replied:
> it's possible

In theory, I don't think this is easy to hit in practice. And #31785 should fix it entirely. You could try modifying `bitcoin-mine` in #30437.

vasild replied:
Ok, I decided to spend time reviewing #31785 instead of trying to trigger this assert. Now, should merging #31785 be a blocker for this PR? If not, then I should try to trigger this assert and confirm that it is not triggered (if this is to be merged without #31785).

Sjors commented on src/node/interfaces.cpp:
```diff
@@ -936,9 +937,90 @@ class BlockTemplateImpl : public BlockTemplate
+                WAIT_LOCK(notifications().m_tip_block_mutex, lock);
+                notifications().m_tip_block_cv.wait_until(lock, std::min(now + tick, deadline), [&]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
+                    Assume(notifications().TipBlock());
+                    tip_changed = notifications().TipBlock() != m_block_template->block.hashPrevBlock;
```
Taken (with some changes)

Sjors commented:
I went back to using the `MAX_MONEY` magic value and added a helper `check_tip_changed()` to make sure we check for a tip update _before_ waiting.

Sjors commented on src/node/interfaces.cpp:
```diff
@@ -936,9 +937,90 @@ class BlockTemplateImpl : public BlockTemplate
+            bool tip_changed{false};
+            {
+                WAIT_LOCK(notifications().m_tip_block_mutex, lock);
+                notifications().m_tip_block_cv.wait_until(lock, std::min(now + tick, deadline), [&]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
```
I added check before we start the wait.

Sjors replied:
Actually, I forgot that `wait_until` already checks the predicate once before waiting. See https://github.com/bitcoin/bitcoin/pull/31283#discussion_r1938970989

Sjors commented on src/node/interfaces.cpp:
```diff
@@ -936,9 +937,90 @@ class BlockTemplateImpl : public BlockTemplate
+                auto block_template{std::make_unique<BlockTemplateImpl>(m_assemble_options, BlockAssembler{chainman().ActiveChainstate(), context()->mempool.get(), m_assemble_options}.CreateNewBlock(), m_node)};
+
+                // If the tip changed, return the new template regardless of its fees.
+                if (tip_changed) {
```
If that happens we'll catch it at the start of the next loop, with the above change.

Sjors replied:
`m_tip_block_mutex` and `cs_main` locks don't jive well, so I limited the places where we actually check

vasild commented on src/node/interfaces.cpp:
```diff
@@ -936,9 +937,112 @@ class BlockTemplateImpl : public BlockTemplate
+        const bool allow_min_difficulty{chainman().GetParams().GetConsensus().fPowAllowMinDifficultyBlocks};
+
+        bool tip_changed{false};
+        // Helper to check if the tip has has changed, and also update tip_changed.
```
nit:
```suggestion
        // Helper to check if the tip has changed, and also update tip_changed.
```

vasild commented on src/node/interfaces.cpp:
```diff
@@ -936,9 +937,112 @@ class BlockTemplateImpl : public BlockTemplate
+        while (now <= deadline) {
+            {
+                WAIT_LOCK(notifications().m_tip_block_mutex, lock);
+                // First make sure the tip hasn't already changed:
+                check_tip_changed() || notifications().m_tip_block_cv.wait_until(lock, std::min(now + tick, deadline), [&]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
+                    return check_tip_changed() || chainman().m_interrupt;
+                });
```
`wait_until()` will first call the predicate, so `check_tip_changed() || ` is not needed. The above code is equivalent to:

```cpp
                notifications().m_tip_block_cv.wait_until(lock, std::min(now + tick, deadline), [&]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
                    return check_tip_changed() || chainman().m_interrupt;
                });
```

Sjors replied:
Ah oops, we're going in circles, I forgot about this as well. Will add a comment.

Sjors replied:
@ryanofsky pointed out the same mistake in a different (now closed) PR: https://github.com/bitcoin/bitcoin/pull/31297#discussion_r1848409060

Sjors commented:
Dropped unnecessary `check_tip_changed()` check before `wait_until`..

Sjors commented:
Applied @vasild's suggestion: https://github.com/bitcoin/bitcoin/pull/31283#discussion_r1939075014

vasild reviewed (approved):
ACK 276ce2eea38865154017047f3761225c2b504cf6

Two things that would be nice to figure out before merge:

* [x] There could be a low-hanging fruit to avoid iterating over all transactions and use `-nFee` which should be significantly faster: https://github.com/bitcoin/bitcoin/pull/31283#discussion_r1937311760

* [ ] Just to confirm that the `Assume()` can't happen in normal circumstances: https://github.com/bitcoin/bitcoin/pull/31283#discussion_r1939240601

Sjors commented:
Rebased after #31384.

ryanofsky commented on src/node/interfaces.cpp:
```diff
@@ -942,9 +943,94 @@ class BlockTemplateImpl : public BlockTemplate
+        // The latter check is expensive so we only run it once per second.
+        auto now{NodeClock::now()};
+        const auto deadline = now + options.timeout;
+        const MillisecondsDouble tick{1000};
```
In commit "Add waitNext() to BlockTemplate interface" (7f6f58bad8e56c9a25ffbc0d105bed37e9958fef)

Any thoughts on whether it makes sense for this polling interval to be configurable? I could imagine a miner not really caring how expensive this check is and just wanting it to be as fast as possible?

Sjors replied:
A miner might naively set it so low that our p2p processing is impacted, potentially getting new blocks later and interfering with transaction relay.

Also, I hope to get rid of this soon(tm) if cluster mempool gives us a more efficient method. It's nice if we don't have to drop a configuration option again.

This is already 60x faster than the long poll behavior of `getblocktemplate`, which nobody complains about. And typical stratum (v1) jobs are refreshed every 30 seconds. https://b10c.me/blog/014-mining-pool-behavior-during-forks/

Sjors replied:
Additionally there's memory management to worry about, if we really start pushing fresh template at every mempool increment, potentially multiple times per second.

ryanofsky commented on src/test/miner_tests.cpp:
```diff
@@ -672,8 +677,10 @@ BOOST_AUTO_TEST_CASE(CreateNewBlock_validity)
+        if (current_height % 2 == 0) {
+            block_template = mining->createNewBlock(options);
+            BOOST_REQUIRE(block_template);
+        } // for odd heights the next template is created at the end of this loop
```
In commit "Add waitNext() to BlockTemplate interface" (7f6f58bad8e56c9a25ffbc0d105bed37e9958fef)

I found this comment pretty confusing and would suggest making it more specific like "// if current_height is odd, block_template will have already been set at the end of the previous loop."

ryanofsky commented on src/test/miner_tests.cpp:
```diff
@@ -710,7 +717,12 @@ BOOST_AUTO_TEST_CASE(CreateNewBlock_validity)
             BOOST_REQUIRE_EQUAL(maybe_new_tip->GetBlockHash(), block.GetHash());
         }
         // This just adds coverage
```
In commit "Add waitNext() to BlockTemplate interface" (7f6f58bad8e56c9a25ffbc0d105bed37e9958fef)

The "// This just adds coverage" comment should be moved above the waitTipChanged call since it does not apply to the waitNext call, which is needed to set `block_template` for the next loop iteration.

ryanofsky commented on src/node/interfaces.cpp:
```diff
@@ -942,9 +943,94 @@ class BlockTemplateImpl : public BlockTemplate
+            const auto tip_block{notifications().TipBlock()};
+            // This is an instance method on BlockTemplate and no template
+            // could have been generated before a tip exists.
+            Assume(tip_block);
+            tip_changed = tip_block != m_block_template->block.hashPrevBlock;
```
In commit "Add waitNext() to BlockTemplate interface" (7f6f58bad8e56c9a25ffbc0d105bed37e9958fef)

Seem like it would be more correct (and more compact) to write this as:

tip_changed = Assume(tip_block) && tip_block != m_block_template->block.hashPrevBlock;

Since if tip_block is null the right thing to do is just wait for it to be non-null, not set tip_changed to true and try to return a new block template below.

vasild replied:
> Assume(tip_block) ... if tip_block is null the right thing to do is just wait ...

Hmm, if `tip_block` is null then `Assume()` will not wait but stop the program?

Sjors replied:
Taken

`Assume(x)` is `x` in production, so in the event of a bug `tip_changed` will be false. So the function will just wait for the deadline and return nothing.

ryanofsky reviewed (approved):
Code review ACK ce9c2e17d5d4ef99cba82521f2515e34130a02a1. I don't think this needs to depend on the other PR #31785, because I don't see how the assume check https://github.com/bitcoin/bitcoin/pull/31283#discussion_r1937596836 could fail. Left some minor comments below though, including one about how assume check could be handled better.

vasild commented on src/node/interfaces.cpp:
```diff
@@ -942,9 +943,94 @@ class BlockTemplateImpl : public BlockTemplate
+
+        bool tip_changed{false};
+        // Helper to check if the tip has changed, and also update tip_changed.
+        auto check_tip_changed = [this, &tip_changed]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
```
Before it was like what you suggest, then the `check_tip_changed()` lambda was introduced because that code had to be called in two places, then, following suggestions, it was changed to be called in just one place. I agree it would be better to remove it now, given that it is called in just one place.

Sjors commented on src/node/types.h:
```diff
@@ -61,6 +63,28 @@ struct BlockCreateOptions {
+     */
+    MillisecondsDouble timeout{MillisecondsDouble::max()};
+
+    /**
+     * Wait until total fees in the new template exceed fees in the origal
+     * template by at least this amount (in sats). The default is to ignore
+     * fee increases and only wait for a tip change.
+     *
+     * This is not an std::optional<CAmount> because (the current version of)
+     * libmultiprocess maps missing values to 0.
+     * https://github.com/bitcoin/bitcoin/pull/31283#discussion_r1937622086
+     *
+     * Use a magic value instead.
+     */
+    CAmount fee_threshold{MAX_MONEY};
```
I incorporated part of this comment. Note that some callers may just not be interested in fee changes, so I mentioned that as well.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/31283#discussion_r1954786986

In commit "Add waitNext() to BlockTemplate interface" (86449088b703d2110edaacbd6d1f1386916a3773)

> I incorporated part of this comment. Note that some callers may just not be interested in fee changes, so I mentioned that as well.

Thanks, I do have two more suggestions:

- In "A caller may not be interested in templates with higher fees, but determining whether fee_threshold is reached is also expensive," can you replace "but" with "and". Both of these are just different reasons to specify `MAX_MONEY`.
- Can we drop the "This is not an std::optional<CAmount> [...]" section? I think as long as the MAX_MONEY case is just an optimization (we could drop rid of the `options.fee_threshold < MAX_MONEY` check in the code and it would be less efficient internally but not have an observable difference from the callers perspective), it doesn't make actually make sense to use std::optional here. Using `std::optional` would just make the interface more confusing and the implementation more complicated.

Sjors commented:
(there seems to be a delay in Github processing my latest push)

ryanofsky reviewed (approved):
Code review ACK 16b18f845313f25bbcdea122baa753b570add7d0. Just various suggested changes since last review.

> (there seems to be a delay in Github processing my latest push)

Can confirm I see "Processing Updates" "Recent push is still being processed, and will show up in a bit" at the top of this PR.

Sjors commented:
It seems Github never around to my last push. The new one was picked up immediately. It incorporates the two suggestions.

ryanofsky reviewed (approved):
Code review ACK 2d65225c4696dcc8312d546eb67f4eae11b9b8fe, just updating fee_threshold comment since last review

ismaelsadeeq commented on src/node/interfaces.cpp:
```diff
@@ -942,9 +943,89 @@ class BlockTemplateImpl : public BlockTemplate
+                // Note that wait_until() checks the predicate before waiting
+                notifications().m_tip_block_cv.wait_until(lock, std::min(now + tick, deadline), [&]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
+                    AssertLockHeld(notifications().m_tip_block_mutex);
+                    if (tip_changed) return true;
```
@Sjors , does `tip_changed` ever become `true` at line 956 and we return from there? I think it will always be false hence can be deleted.

1. When we first enter this scope, the predicate is executed before blocking. If `tip_changed` is updated to `true`, it is returned. And when the predicate returned value is `true`, we do not wait; instead, we proceed to build and return a block template.

2. In subsequent executions of the predicate `tip_changed` will be `false` and we will check if the tip has changed, if yes `tip_changed` is updated to `true`, and returned, causing the wait to end. We then proceed to build and return a block template.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/31283/files#r1960644463

Agree it looks like this line can be deleted. Also scope of `tip_changed` can be reduced so it is inside the do..while loop.

Sjors replied:
Done both.

ismaelsadeeq commented on src/node/interfaces.cpp:
```diff
@@ -942,9 +943,89 @@ class BlockTemplateImpl : public BlockTemplate
+             * (approximate) fees for the next block increased, perhaps more so after
+             * Cluster Mempool.
+             *
+             * We'll also create a new template if the tip changed during the last tick.
```
We can generate a template before a tick

vasild commented on src/node/interfaces.cpp:
```diff
@@ -942,9 +943,98 @@ class BlockTemplateImpl : public BlockTemplate
+                    // method on BlockTemplate and no template could have been
+                    // generated before a tip exists.
+                    tip_changed = Assume(tip_block) && tip_block != m_block_template->block.hashPrevBlock;
+                    return tip_changed;
```
This was `return check_tip_changed() || chainman().m_interrupt;` when the lambda `check_tip_changed()` was used. Now that the lambda is expanded here, the `chainman().m_interrupt` condition was lost. I think it should be:

```suggestion
                    return tip_changed || chainman().m_interrupt;
```

Sjors replied:
Fixed

vasild reviewed (approved):
ACK 2d65225c4696dcc8312d546eb67f4eae11b9b8fe modulo the comment below and https://github.com/bitcoin/bitcoin/pull/31283#discussion_r1960644463

Sjors commented on src/node/interfaces.cpp:
```diff
@@ -877,7 +878,7 @@ class ChainImpl : public Chain
 {
 public:
-    explicit BlockTemplateImpl(std::unique_ptr<CBlockTemplate> block_template, NodeContext& node) : m_block_template(std::move(block_template)), m_node(node)
+    explicit BlockTemplateImpl(BlockAssembler::Options assemble_options, std::unique_ptr<CBlockTemplate> block_template, NodeContext& node) : m_assemble_options(std::move(assemble_options)), m_block_template(std::move(block_template)), m_node(node)
```
`clang-format` doesn't seem to touch this. However once I added some line breaks myself, it suddenly had an opinion. So I'll commit that.

vasild replied:
`clang-format` will not break long lines itself because `src/.clang-format` contains `ColumnLimit: 0` (unlimited). However, if you break the line yourself it will do the alignment according to the other rulez.

I have a `clang-format` wrapper shell script that replaces `ColumnLimit: 0` with `ColumnLimit: 110` in `src/.clang-format`, runs the actual `clang-format` and then restores the file (so that I do not commit it accidentally).

Sjors commented:
Rebased and addressed feedback. I also reformatted the `BlockAssembler{` call to make it a bit less wide.

ryanofsky reviewed (approved):
Code review ACK cadbd4137d84b71be26effd6a2ae177d5031345e. Main change since last review is adding back a missing `m_interrupt` check in the waitNext loop. Also made various code cleanups in both commits.

ryanofsky commented on src/node/interfaces.cpp:
```diff
@@ -942,9 +947,94 @@ class BlockTemplateImpl : public BlockTemplate
         return chainman().ProcessNewBlock(block_ptr, /*force_processing=*/true, /*min_pow_checked=*/true, /*new_block=*/nullptr);
     }
 
+    std::unique_ptr<BlockTemplate> waitNext(BlockWaitOptions options) override
```
> As far as I can see the current one becomes useless, as its `m_block_template` will be outdated, so future calls to `waitNext` for it will return instantly anyway.

This should not be the case. This returns a pointer to a new block template so miners are able to mine with the old one while calling this.

ryanofsky replied:
> Sure, while calling `waitNext`, but after it returns, isn't it the case that the current `BlockTemplate` becomes useless?

It's not useless if submitSolution can still be called. This should be the case as long as the tip hasn't changed and miners are still mining with this template. If the tip has changed or miners are no longer mining with this template, the client can discard it.

ismaelsadeeq reviewed (approved):
Code review ACK cadbd4137d84b71be26effd6a2ae177d5031345e

Sjors commented on src/node/interfaces.cpp:
```diff
@@ -942,9 +947,94 @@ class BlockTemplateImpl : public BlockTemplate
+        // comes in before the next tick.
+        CAmount current_fees = -1;
+
+        // Alternate waiting for a new tip and checking if fees have risen.
+        // The latter check is expensive so we only run it once per second.
```
The phrasing is a bit unclear, but "alternate" refers to what happens within each tick: we first check if the tip changed, then check if the fees went up.

vasild reviewed (approved):
ACK cadbd4137d84b71be26effd6a2ae177d5031345e

ryanofsky commented:
Looks like this has 3 acks so should be good to merge after the 29 release branch is created (https://github.com/bitcoin/bitcoin/issues/31029)

Sjors commented:
@ryanofsky what would be a good way for a consumer of `waitNext()`, directly or via IPC, to abort a wait? Currently only a node shutdown can interrupt it, but in the context of a Template Provider, a stratum client disconnect should probably also abort the wait.

Sjors commented:
@ryanofsky thanks. I'm probably going to leave that to followup, to preserve ACKs and limit the scope of this PR.

vasild commented:
Or could pass an additional argument to `waitNext()`: `CThreadInterrupt` and have `waitNext()` honor that one in addition to `chainman().m_interrupt`.

Sjors commented:
The problem is a bit less urgent now, since I refactored the Template Provider to have one thread per connection. Worst case this thread lingers for half an hour until a new block arrives. It only becomes a potential DoS problem for public facing template providers, but that use case needs additional safety and performance changes anyway.
